autogpt/memory/sqlitemem.py

```diff
@@ -0,0 +1,368 @@
+    ) -> None:
+        """
+        Converts the embeddings from a different embedder to the current one.
+
+        Args:
+            old_embeddings: The path to the previous embeddings
+            prev_embedder: The name of the previous embedder
+
+        Returns: None
+        """
+
+        # Ask user if they want to regenerate embeddings (may come with additional costs for embedders like "ada")
+        print(
+            f"WARNING: Embeddings for index '{self.index}' were generated "
+            f"with a different embedder ({prev_embedder}) than the current one ({self.embedder_name}). "
+            "Regenerating embeddings (may come with additional costs if using embedders like 'ada'), "
+            "do you want to regenerate them [y/N]?",
+            end="",
+        )
+        if input().lower() != "y":
+            logger.error(
+                "Embeddings for this index were generated with a different embedder. "
+                "Either regenerate them, use a different embedder or remove the old embeddings file."
+            )
+            exit(1)
+
+        # Create the new embeddings array
+        self._create_embeddings()
+
+        # Get the data from the database
+        with self.sql_conn as conn:
+            res = conn.execute(f"SELECT id, data, timestamp FROM {self.index};")
+            rows = res.fetchall()
+
+        # For each data point, get the embedding and add it to the new embeddings array
+        for index, data, timestamp in rows:
+            embedding = get_ada_embedding(data)
+            embedding = np.expand_dims(embedding, axis=0)
+            self.embeddings[index] = embedding
+
+        # Rename the old embeddings
+        if soft_delete:
+            old_embeddings.rename(old_embeddings.with_suffix(".zarr.old"))
+        else:
+            shutil.rmtree(old_embeddings, ignore_errors=True)
+
+    def add(self, data: str) -> str:
+        """
+        Adds a data point to the memory.
+
+        Args:
+            data: The data to add.
+
+        Returns: Message indicating that the data has been added.
+        """
+        # Create the data embedding
+        embedding = get_ada_embedding(data)
+        embedding = np.expand_dims(embedding, axis=0)
+
+        # Get the embedding index (index of the next vector is the same as the number of vectors)
+        embedding_index = self.embeddings.shape[0]
+
+        # Add data to the database
+        with self.sql_conn as conn:
+            try:
+                conn.execute(
+                    f"INSERT INTO {self.index} (id, data, timestamp) VALUES(?, ?, CURRENT_TIMESTAMP);",
+                    (embedding_index, data),
+                )
+            except sqlite3.IntegrityError as err:
+                if "UNIQUE constraint failed:" in str(err):
+                    return "Trying to add duplicate data to memory, ignoring."
+                raise err
+
+        # Store the embedding
+        self.embeddings.append(embedding)
```

Comment:
If all of the embeddings are stored in numpy what is the point of storing them in sql? This seems like you will be memory bound by the number of embeddings you can have. 
Shouldn't either everything be in memory or in sql but not both?
It seems like this just saves a compressed representation to disk, but the whole uncompressed embedding lives in memory